internals/src/error.rs

```diff
@@ -26,3 +26,31 @@ macro_rules! write_err {
         }
     }
 }
+
+/// This module is used internally to check that error types implement certain traits.
+#[cfg(feature = "trait-checks")]
+#[doc(hidden)]
+pub mod trait_checks {
+    /// Internal trait to assert that error types implement `Send` and `Sync`.
+    pub trait AssertSendSync: Send + Sync {}
+
+    /// Internal trait to assert that error types implement `std::error::Error` (implies `Display` + `Debug`).
+    #[cfg(feature = "std")]
+    pub trait AssertError: std::error::Error {}
+
+    /// Checks at compile time that type `$t` implements the traits we require for error types.
+    #[macro_export]
+    macro_rules! check_pub_error {
+        ($ty:ident) => {
+            $crate::check_pub_error!($ty, );
+        };
+        ($ty:ident, $($gen:ident),*) => {
+            use internals::error::trait_checks::*;
+
+            impl<$($gen),*> AssertSendSync for $ty<$($gen),*> where $($gen: Send + Sync ),* {}
+
+            #[cfg(feature = "std")]
+            impl<$($gen),*> AssertError for $ty<$($gen),*> where $($gen: std::error::Error + 'static),* {}
```

Comment:
Could also check `Debug + Display` for `no_std`.

Reply:
Done as a separate trait.